keras_cv/bounding_box/non_max_suppression.py

```diff
@@ -0,0 +1,181 @@
+                            ],
+    ], dtype = np.float32)
+
+    nms = NonMaxSuppression(
+        num_classes=8,
+        bounding_box_format="center_xyWH",
+        iou_threshold=0.1
+    )
+
+    boxes = nms(boxes, images)
+    ```
+    """
+
+    def __init__(
+        self,
+        num_classes,
+        bounding_box_format,
+        confidence_threshold=0.05,
+        iou_threshold=0.5,
+        max_detections=100,
+        max_detections_per_class=100,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.num_classes = num_classes
+        self.bounding_box_format = bounding_box_format
+        self.confidence_threshold = confidence_threshold
+        self.iou_threshold = iou_threshold
+        self.max_detections = max_detections
+        self.max_detections_per_class = max_detections_per_class
+
+    def call(self, predictions, images=None):
+        # convert to yxyx for the TF NMS operation
+        predictions = convert_format(
+            predictions,
+            source=self.bounding_box_format,
+            target="yxyx",
+            images=images,
+        )
+
+        # preparing the predictions for TF NMS op
+        boxes = tf.expand_dims(predictions[..., :4], axis=2)
+        classes = tf.cast(predictions[..., 4], tf.int32)
+        scores = predictions[..., 5]
+
+        classes = tf.one_hot(classes, self.num_classes)
+        scores = tf.expand_dims(scores, axis=-1) * classes
+
+        # applying the NMS operation
+        nmsed_boxes = tf.image.combined_non_max_suppression(
+            boxes,
+            scores,
+            self.max_detections_per_class,
+            self.max_detections,
+            self.iou_threshold,
+            self.confidence_threshold,
+            clip_boxes=False,
+        )
+
+        # output will be a ragged tensor because num_boxes will change across the batch
+        return self._encode_to_ragged(nmsed_boxes, images)
+
+    def _encode_to_ragged(self, nmsed_boxes, images):
+        # this TensorArray will hold all the valid detections
+        boxes = tf.TensorArray(tf.float32, size=0, dynamic_size=True)
+
+        for i in tf.range(tf.shape(nmsed_boxes.nmsed_boxes)[0]):
+            num_detections = nmsed_boxes.valid_detections[i]
+
+            # recombining with classes and scores
+            boxes_recombined = tf.concat(
+                [
+                    nmsed_boxes.nmsed_boxes[i][:num_detections],
+                    tf.expand_dims(
+                        nmsed_boxes.nmsed_classes[i][:num_detections], axis=-1
+                    ),
+                    tf.expand_dims(
+                        nmsed_boxes.nmsed_scores[i][:num_detections], axis=-1
+                    ),
+                ],
+                axis=-1,
+            )
+
+            # iterate through the boxes and append it to TensorArray
+            for i in range(nmsed_boxes.valid_detections[i]):
+                boxes = boxes.write(boxes.size(), boxes_recombined[i])
```

Comment:
you're shadowing i here.

for i in range(...[i]).

You need to use another variable, like j.

Then I think you can replace boxes.size() with j.

Reply:
That's a great catch, thanks!

That being said, I have left the `boxes.size()` as it is to mimic the `append` function whereas `j` will reset for every batch.